readers/multishard.cc

```diff
@@ -820,10 +823,34 @@ future<> shard_reader_v2::do_fill_buffer() {
                 });
                 auto s = gs.get();
                 auto permit = co_await _lifecycle_policy->obtain_reader_permit(s, "shard-reader", timeout());
-                auto rreader = make_foreign(std::make_unique<evictable_reader_v2>(evictable_reader_v2::auto_pause::yes, std::move(ms),
-                            s, std::move(permit), *_pr, _ps, _pc, _trace_state, _fwd_mr));
+                if (permit.needs_readmission()) {
+                    co_await permit.wait_readmission();
+                }
+                auto underlying_reader = _lifecycle_policy->create_reader(s, permit, *_pr, _ps, _pc, _trace_state, _fwd_mr);
 
                 std::exception_ptr ex;
+
+                try {
+                    // The reader might have been saved from a previous page and
+                    // missed some fast-forwarding since the new page started.
+                    // Fast forward it to the correct range if that is the case.
```

Comment:
How could this have happened? If we missed a page, shouldn't we also miss in the querier cache and result in a completely new reader being created?

Reply:
We didn't miss a page. The reader might have missed a range, because e.g. it was completely served from other shards. By the time the reads gets to this shard the first time, we might be on another range, not the first one.

Reply:
Aha, understood.

Reply:
Now looking at it from a different perspective, perhaps we should change the multishard reader's fast-forward to talk to fast-forward all of the shards. Or even to just evict all the caches.

The reasoning is that it's safer to do the right thing up-front than to correct things after the fact. The correction might happen due to some other bug, and then we'd never know about it.

But, we already do that:

```c++
future<> multishard_combining_reader_v2::fast_forward_to(const dht::partition_range& pr) {
    clear_buffer();
    _end_of_stream = false;
    on_partition_range_change(pr);
    return parallel_for_each(_shard_readers, [&pr] (std::unique_ptr<shard_reader_v2>& sr) {
        return sr->fast_forward_to(pr);
    });
}
```

So why the bug? Maybe the fix is legitimate, but is the comment correct? The reader cannot have missed a fast forward operation, or if it could, how?

Reply:
The shard reader skips the `fast_forward_to()` if the reader is not created yet. But there is an abstraction leak here: the shard reader doesn't know this, but the *uncreated* reader might be a reader sitting in the querier cache. When the shard reader later creates the underlying reader, it will get a reader which is not at the range that it passed to the mutation source. The mutation source which returns this reader could correct the range of the returned reader, but it can't because mutation source is synchronous but `fast_forward()` isn't.
So we are left with the workaround introduced in this patch.

I also had a version where I evict the the cached reader on fast-forward-to but that blew up in my face (don't remember why though).